python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +79,102 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
 
         """
         super().setP(P, Pmat)
+
+
+class SNESSolver:
```

Comment:
Why wrap SNESSolver? We don't do this for e.g. KSP.

Reply:
We wrap `KSP` in `LinearProblem`.

Reply:
It also matches how we have split the `NewtonSolver` into two instances. You have a pure solver class, doing the algorithmic thing, getting information from  `NonLinearProblem`

Reply:
I don't like this split ala legacy code into a FooSolver and FooProblem classes. I never understood why this is needed and always confused me. I think of one FooProblem class that has the algorithmic active `.solve()` etc methods.

Reply:
I'm fine with merging them (also in `NonLinearProblem`/`NewtonSolver` if everyone else is fine with it.
Seems like a history reason (@garth-wells can probably give some context: https://github.com/FEniCS/dolfinx/blob/089c89b1e3746e6b875d98a2a31ebc2775e191a4/src/kernel/nls/NonlinearProblem.cpp) 
which in turn was mirrored in: https://github.com/FEniCS/dolfinx/pull/1478

Reply:
The C++ Newton solver was never properly reviewed with dolfin -> dolfinx. It shouldn't necessarily be taken as a good design model to follow.

Reply:
Factory function has been added here. There is now only one solver class, that accepts any kind of input that follows the typing protocol above. This makes it as general as I can imagine.